packages/server/postgres/utils/checkEqBase.ts

```diff
@@ -0,0 +1,164 @@
+import {RTable, TableSchema} from '../../database/stricterR'
+import areEqual from 'fbjs/lib/areEqual'
+import getPg from '../getPg'
+
+interface DBDoc {
+  id: string
+  [key: string]: any
+}
+interface RethinkDoc extends DBDoc {}
+interface PGDoc extends DBDoc {}
+
+interface IError {
+  [key: string]:
+    | {
+        error: string[]
+        rethinkRow: Partial<RethinkDoc>
+        pgRow: Partial<PGDoc>
+      }
+    | number
+    | undefined
+    | string[]
+  rethinkRecordsCompared: number
+  foundErrors: number
+  missingPgRows: string[]
+  extraPgRows: string[]
+  rethinkRowCount?: number
+  pgRowCount?: number
+}
+
+function getPairNeFields(
+  rethinkRow: RethinkDoc,
+  pgRow: PGDoc,
+  alwaysDefinedFields: string[],
+  maybeUndefinedFieldsDefaultValues: {[key: string]: any}
+): string[] {
+  const neFields = [] as string[]
+
+  for (const f of alwaysDefinedFields) {
+    const [rethinkValue, pgValue] = [rethinkRow[f], pgRow[f]]
+    if (!areEqual(rethinkValue, pgValue)) {
+      neFields.push(f)
+    }
+  }
+  for (const [f, defaultValue] of Object.entries(maybeUndefinedFieldsDefaultValues)) {
+    const [rethinkValue, pgValue] = [rethinkRow[f], pgRow[f]]
+    if (rethinkValue !== undefined) {
+      if (!areEqual(rethinkValue, pgValue)) {
+        neFields.push(f)
+      }
+    } else {
+      if (!areEqual(pgValue, defaultValue)) {
+        neFields.push(f)
+      }
+    }
+  }
+
+  return neFields
+}
+
+function addNeFieldsToErrors(
+  errors: IError,
+  neFields: string[],
+  rethinkRow: RethinkDoc,
+  pgRow: PGDoc,
+  rowId: string
+): void {
+  const rethinkNeFields = {}
+  const pgNeFields = {}
+  for (const neField of neFields) {
+    rethinkNeFields[neField] = rethinkRow[neField]
+    pgNeFields[neField] = pgRow[neField]
+  }
+  errors[rowId] = {
+    error: neFields,
+    rethinkRow: rethinkNeFields,
+    pgRow: pgNeFields
+  }
+}
+
+const getExtraPgRowIds = async (tableName: string): Promise<string[]> => {
+  const pg = getPg()
+  const queryRes = await pg.query(`
+    SELECT ARRAY_AGG("id") FROM "${tableName}"
+      WHERE "eqChecked" = '-infinity'::timestamptz;
+  `)
+  return queryRes.rows[0].array_agg
+}
+
+export async function checkTableEq(
+  tableName: string,
+  rethinkQuery: RTable<TableSchema>,
+  pgQuery: (
+    update: {
+      [key: string]: any
+      eqChecked: Date
+    },
+    ids: string[]
+  ) => Promise<PGDoc[]>,
+  alwaysDefinedFields: string[],
+  maybeUndefinedFieldsDefaultValues: {[key: string]: any},
+  maxErrors = 10
+): Promise<IError> {
+  const errors: IError = {
+    rethinkRecordsCompared: 0,
+    foundErrors: 0,
+    missingPgRows: [],
+    extraPgRows: []
+  }
+  const batchSize = 3000
+
+  for (let i = 0; i < 1e5; i++) {
+    if (errors.foundErrors === maxErrors) {
```

Comment:
nit: i always try to write `>=` in these cases because gremlins will creep into my code, cause a float error & increment by 1.00000001 when i least expect it (doesn't seem to be the case here)

Reply:
yea this is a good point